data_model/src/events/data/filters.rs

```diff
@@ -1,192 +1,734 @@
+        Serialize,
+        IntoSchema,
+    )]
+    pub enum DataEventFilter {
+        /// Matches any data events ([`DataEvent`])
+        Any,
+        /// Matches [`PeerEvent`]s
+        Peer(PeerEventFilter),
+        /// Matches [`DomainEvent`]s
+        Domain(DomainEventFilter),
+        /// Matches [`AccountEvent`]s
+        Account(AccountEventFilter),
+        /// Matches [`AssetEvent`]s
+        Asset(AssetEventFilter),
+        /// Matches [`AssetDefinitionEvent`]s
+        AssetDefinition(AssetDefinitionEventFilter),
+        /// Matches [`TriggerEvent`]s
+        Trigger(TriggerEventFilter),
+        /// Matches [`RoleEvent`]s
+        Role(RoleEventFilter),
+        /// Matches [`PermissionTokenSchemaUpdateEvent`]s
+        // nothing to filter for, really
+        PermissionTokenSchemaUpdate,
+        /// Matches [`ConfigurationEvent`]s
+        Configuration(ConfigurationEventFilter),
+        /// Matches [`ExecutorEvent`]s
+        Executor(ExecutorEventFilter),
+    }
+
+    /// An event filter for [`PeerEvent`]s
+    #[derive(
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Getters,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    pub struct PeerEventFilter {
+        /// If specified matches only events originating from this peer
+        pub(super) id_matcher: Option<super::PeerId>,
+        /// Matches only event from this set
+        pub(super) event_set: PeerEventSet,
+    }
+
+    /// An event filter for [`DomainEvent`]s
+    #[derive(
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Getters,
         Decode,
         Encode,
         Deserialize,
         Serialize,
         IntoSchema,
     )]
-    #[serde(untagged)] // Unaffected by #3330
-    pub enum FilterOpt<F> {
-        /// Accept all items that will be passed to `filter()` method
-        #[serde(with = "accept_all_as_string")]
-        AcceptAll,
-        /// Use filter `F` to choose acceptable items passed to `filter()` method
-        BySome(F),
+    pub struct DomainEventFilter {
+        /// If specified matches only events originating from this domain
+        pub(super) id_matcher: Option<super::DomainId>,
+        /// Matches only event from this set
+        pub(super) event_set: DomainEventSet,
     }
 
+    /// An event filter for [`AccountEvent`]s
     #[derive(
-        Debug, Clone, PartialEq, Eq, FromVariant, Decode, Encode, Deserialize, Serialize, IntoSchema,
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Getters,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    pub struct AccountEventFilter {
+        /// If specified matches only events originating from this account
+        pub(super) id_matcher: Option<super::AccountId>,
```

Comment:
I think it would increase "blindness". If we seek for shorter names, this would work better imo:

```rust
struct Matcher<T>(Option<T>);

pub struct AccountEventFilter {
  id: Matcher<super::AccountId>
}
```

Reply:
Actually, I am thinking of changing it from `Option` to an actual `AccountIdMatcher` that will have some additional logic for matching by domain id, for example, which would remove the ADT blindness concern.

Before that I would prefer to leave the name as-is..